### DISCUSSION TITLE
How can define a dbt source as a Dagster asset?

### DISCUSSION THREAD

**Discussion by marcilj** [dagster-io/dagster#17648] [Q&A] [answered]
I'm trying to understand how can I configure DBT Sources as Dagster assets following the documentation provided [here](https://docs.dagster.io/integrations/dbt/reference#defining-a-dbt-source-as-a-dagster-asset).

The part that I don't understand is the return or yield part.

In my situation I have a table in Snowflake that is getting updated `without` Dagster. My goal is to create a Dagster asset for that DBT source, so I can add description, generate asset observation, create Data Version function and trigger run based on the presence of new data in the source, etc.

I figure that for this I would need to define the source as an `@asset` using `get_asset_key_for_source()` function or to define a `SourceAsset` and select them source from the `key`.

If I configure the asset using this code

```yml
sources:
  - name: test
```

```python
from dagster import asset, AssetExecutionContext
from dagster_dbt import DbtCliResource, get_asset_key_for_source, dbt_assets

@dbt_assets(
    manifest=dbt_manifest_path,
    select="tag:my_model_tag",
)
def my_dbt_asset(
    context: OpExecutionContext,
    dbt: DbtCliResource,
    config: DbtConfig,
):
    yield from dbt.cli("run", context=context).stream()

@asset(key=get_asset_key_for_source([my_dbt_assets], "test"))
def my_dbt_source_as_dagster_asset():
    return ...
```

What does the return part of `my_dbt_source_as_dagster_asset` asset look like?

I could create a Snowflake Client, and then execute the query `select * from my_dbt_source_as_dagster_asset;` and return the value, but that doesn't seems right.

Then I'm thinking of using the Snowflake IO Manager instead since it's already configure I could do something like 
```python
@asset(
      io_manager_key="s3_io_manager",
      key=get_asset_key_for_source([my_dbt_assets], "test")
)
def my_dbt_source_as_dagster_asset():
    return ...
```
But then again what do I put in my return clause?

I might be thinking this the wrong way, but I can't figure out what goes there since it's basically only pointing to a table in snowflake.

Am I missing something here?
If I do something like `return 1` does that mean, my dbt core that query this source will get the value `1` instead querying the actual table?
I'm lost...

**Comment by sryza** [comment, marked as answer]
Hi @marcilj - the `@asset` decorator is used to define assets that can be materialized within Dagster. If I understand correctly, the asset that you want your dbt models to depend on will not be able to be materialized within Dagster, so I would not use the `@asset` decorator.

If you just want to add a description, you can use the `SourceAsset` decorator:

```python
my_dbt_source_as_dagster_asset = SourceAsset(
    key=get_asset_key_for_source([my_dbt_assets], "test"),
    description="my description"
)
```

If you want to generate asset observations, you can use an observable source asset:

```python
@observable_source_asset(
    name=get_asset_key_for_source([my_dbt_assets], "test")[-1],
    key_prefix=get_asset_key_for_source([my_dbt_assets], "test")[:-1],
    description="my description",
)
def my_dbt_source_as_dagster_asset():
    return DataVersion(get_latest_version(...))
```

Refer to the code snippet at the bottom of these docs for how to put this observable source asset on a schedule.

Btw, it's clunky that you need to set both `name` and `key_prefix` in the latter case. I just posted a PR to address this: https://github.com/dagster-io/dagster/pull/17652. After that PR is merged, you'll be able to do
```python
@observable_source_asset(
    key=get_asset_key_for_source([my_dbt_assets], "test"),
    description="my description",
)
...
```
I will try to include it in next week's release.

**Reply by marcilj** [reply]
Thank you very much @sryza

This is exactly what I was looking for.
